token-lending/program/src/state/reserve.rs

```diff
@@ -565,50 +631,67 @@ impl ReserveFees {
     /// Calculate the owner and host fees on borrow
     pub fn calculate_borrow_fees(
         &self,
-        collateral_amount: u64,
+        borrow_amount: Decimal,
+        fee_calculation: FeeCalculation,
     ) -> Result<(u64, u64), ProgramError> {
         let borrow_fee_rate = Rate::from_scaled_val(self.borrow_fee_wad);
         let host_fee_rate = Rate::from_percent(self.host_fee_percentage);
-        if borrow_fee_rate > Rate::zero() && collateral_amount > 0 {
+        if borrow_fee_rate > Rate::zero() && borrow_amount > Decimal::zero() {
             let need_to_assess_host_fee = host_fee_rate > Rate::zero();
             let minimum_fee = if need_to_assess_host_fee {
                 2 // 1 token to owner, 1 to host
             } else {
                 1 // 1 token to owner, nothing else
             };
 
-            let borrow_fee = borrow_fee_rate
-                .try_mul(collateral_amount)?
-                .try_round_u64()?
-                .max(minimum_fee);
+            let borrow_fee_amount = match fee_calculation {
+                FeeCalculation::Exclusive => borrow_amount.try_mul(borrow_fee_rate)?,
+                FeeCalculation::Inclusive => {
+                    let borrow_fee_rate =
+                        borrow_fee_rate.try_div(borrow_fee_rate.try_add(Rate::one())?)?;
+                    borrow_amount.try_mul(borrow_fee_rate)?
+                }
+            };
+
+            let borrow_fee = borrow_fee_amount.try_round_u64()?.max(minimum_fee);
+            if Decimal::from(borrow_fee) >= borrow_amount {
+                msg!("Borrow amount is too small to receive liquidity after fees");
+                return Err(LendingError::BorrowTooSmall.into());
+            }
 
             let host_fee = if need_to_assess_host_fee {
                 host_fee_rate.try_mul(borrow_fee)?.try_round_u64()?.max(1)
             } else {
                 0
             };
 
-            if borrow_fee >= collateral_amount {
-                Err(LendingError::BorrowTooSmall.into())
-            } else {
-                Ok((borrow_fee, host_fee))
-            }
+            Ok((borrow_fee, host_fee))
         } else {
             Ok((0, 0))
         }
     }
 }
 
+/// Calculate fees exlusive or inclusive of an amount
+pub enum FeeCalculation {
+    /// Fee added to amount: fee = rate * amount
+    Exclusive,
+    /// Fee included in amount: fee = (rate / (1 + rate)) * amount
+    Inclusive,
+}
+
 impl Sealed for Reserve {}
 impl IsInitialized for Reserve {
     fn is_initialized(&self) -> bool {
         self.version != UNINITIALIZED_VERSION
     }
 }
 
-const RESERVE_LEN: usize = 602;
+// @TODO: Adjust padding, but what's a reasonable number?
+//        Or should there be no padding to save space, but we need account resizing implemented?
```

Comment:
I don't think we need much here, 256 seems plenty. Account resizing might not be implemented for awhile, hopefully soon though